tslearn/nonmyopic.py

```diff
@@ -0,0 +1,509 @@
+            the number of the cluster for which the distance is wanted
+
+        Returns
+        -------
+        Float: the expected euclidean distance
+        """
+
+        euclidean_dist = np.linalg.norm(
+            X - self.cluster_.cluster_centers_[wanted_cluster, : len(X)]
+        )
+
+        return euclidean_dist
+
+    def get_cluster_prob(self, X):
+        """
+        This function computes the probabilities of the time series xt to be in a cluster of the model
+
+        Parameters
+        ----------
+        X: vector
+            a time series that may be truncated
+
+        Returns
+        -------
+        vector : the probabilities of the given time series to be in the clusters of the model
+        """
+        average_distance, minimum_distance = self.get_avg_dist(X)
+        minimum_distance_to_average = (
+            average_distance - minimum_distance
+        ) / average_distance
+        sum_sk = 0
+        sk = []
+        for current_cluster in range(0, self.number_cluster):
+            d_k = self.get_dist(X, current_cluster)
+            Delta_k = (average_distance - d_k) / average_distance
+            proba_current_cluster = 1 / (
+                math.exp(self.lamb * minimum_distance_to_average)
+                + math.exp(self.lamb * (minimum_distance_to_average - Delta_k))
+                + 10e-6
+            )
+            sk.append(proba_current_cluster)
+            sum_sk = sum_sk + sk[-1]
+        final_sk = [(x / sum_sk) for x in sk]
+        return final_sk
+
+    def cost_function(self, X, tau):
+        """
+        From a incomplete series xt, this function compute the expected cost of a prediction made at time "last time of
+        xt + tau"
+
+        Parameters
+        ----------
+        X: vector
+            a time series that may be truncated for which we want to have the cost
+        tau: int
+            gives the future time for which the cost is calculated
+
+        Returns
+        --------
+        float : the computed cost
+        """
+        cost = 0
+        proba_clusters = self.get_cluster_prob(X=X)
+        truncated_t = X.shape[-1]
+        for current_cluster in range(0, self.number_cluster):
+            cost_y = 0
+            for y in range(0, self.number_classes_):
+                cost_y_hat = 0
+                for y_hat in range(0, self.number_classes_):
+                    if y != y_hat:
+                        cost_y_hat = (
+                            cost_y_hat
+                            + self.pyhatyck_["pyhatycks{0}".format(current_cluster)][
+                                truncated_t + tau - self.minimum_time_stamp
+                            ][y][y_hat]
+                        )
+                cost_y = cost_y + self.pyck_[y, current_cluster] * cost_y_hat
+            cost = cost + proba_clusters[current_cluster] * cost_y
+        cost = cost + self.cost_time(t=truncated_t + tau)
+
+        return cost
+
+    @staticmethod
```

Comment:
This one should probably not be static, and then `function_to_minimize` would not be a required argument anymore (accessible through `self.cost_function`).

Also the name of this function should be more explicit (something like `expected_optimal_prediction_time`).